`gh agent-task view`: add support for `--log` and `--follow`

BagToad opened the pull request:
## Description

This adds support for `--log` and `--follow` to `gh agent-task view`.

## Markdown Rendering by Content Type

This adds support for rendering tool call content according to the content type.

<details><summary>Examples</summary>
<p>

### Plain rendering file lists and repository content

<img width="167" height="104" alt="image" src="https://github.com/user-attachments/assets/00aef489-1560-48d5-ad3b-d69e06b7b9b3" />

### Bash commands rendered alongside the output

<img width="744" height="257" alt="image" src="https://github.com/user-attachments/assets/7f54a0d2-c626-4d64-b456-9d82f497005e" />

### Rendering "viewed" markdown files

<img width="172" height="107" alt="image" src="https://github.com/user-attachments/assets/bfc79ea9-72a9-4572-b2ac-5199917bcce5" />

### Syntax highlighting for "viewed" files

<img width="782" height="583" alt="image" src="https://github.com/user-attachments/assets/a7340b63-03c7-4566-8754-cbb2b0c2d418" />

### Diff rendering for "edited" files

<img width="883" height="401" alt="image" src="https://github.com/user-attachments/assets/79361ff2-5af4-4f7a-a20d-4a1d0d318ca5" />

### Rendering tool call inputs and outputs

<img width="493" height="442" alt="image" src="https://github.com/user-attachments/assets/8a5539b4-54a4-406a-a6d9-f76453494d17" />


</p>
</details> 

## Acceptance Criteria

**Given** I have a session ID to view (a1234b-a1234b) attached to PR 1234
**Given** My CWD is repo foo/bar
**When** I run `gh agent-task view a1234b-a1234b --log`
**Then** A plaintext log is printed to stdout and then the program exits

**Given** I have a session ID to view (a1234b-a1234b) attached to PR 1234
**Given** My CWD is repo foo/bar
**When** I run `gh agent-task view a1234b-a1234b --log --follow`
**Then** All the events up to the latest are printed
**Then** The program continues to poll for logs every 5 seconds, printing new events, and
**When** the session reaches any status except `in_progress`
**Then** the program exits

babakks committed:
feat(agent-task/capi): add `GetSessionLogs` method

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
fix(agent-task/shared): add log renderer

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
feat(agent-task view): add `--log` and `--follow` flags

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
feat(agent-task view): add `--log` and `--follow` flags

Signed-off-by: Babak K. Shandiz <babakks@github.com>

BagToad committed:
Various log rendering improvements

Updated the LogRenderer interface and implementations to accept *iostreams.IOStreams instead of *iostreams.ColorScheme, enabling access to terminal theme and width for improved markdown rendering. Refactored related code, tests, and mocks to support this change, and enhanced log rendering to better handle markdown and code output for various tool calls.

BagToad committed:
Refactor JSON content rendering in log entries

Extracted JSON content rendering into a new helper function renderContentAsJSONMarkdown for reuse. Updated 'report_progress' and 'create' tool call handling to use this helper, improving code clarity and reducing duplication. Also added support for rendering PR descriptions in 'report_progress' tool calls.

BagToad committed:
Improve log rendering and tool call handling

Enhances log rendering by stripping diff formatting from viewed file content and improving JSON rendering with optional labels. Expands the list of recognized tool calls with descriptive titles, especially for Playwright and GitHub MCP server tools. Refactors related helper functions for clarity and robustness.

BagToad committed:
Refactor markdown rendering helper function names

Renamed renderMarkdownWithPadding to renderMarkdownWithFormat and related variables for clarity. Updated comments and TODOs for better context on GUI support and code behavior.

BagToad committed:
Rename relativePath to relativeFilePath in log rendering

Refactored the function name from relativePath to relativeFilePath for clarity and updated all usages accordingly in log rendering functions. Also improved comments in stripDiffFormat for better context.

BagToad committed:
Rename renderToolCall to renderToolCallTitle

Refactors the function renderToolCall to renderToolCallTitle and updates all its usages for clarity. This improves function naming to better reflect its purpose of rendering tool call titles.

BagToad committed:
Update log test expectation files

babakks commented on pkg/cmd/agent-task/shared/log_test.go:
```diff
@@ -0,0 +1,64 @@
+			want, err := os.ReadFile(tt.want)
+			require.NoError(t, err)
+
+			// // Temp for updating tests
+			// os.WriteFile(tt.log+".got", stdout.Bytes(), 0644)
```
Leftover from my work.

BagToad replied:
I feel like these are fine to leave at the moment because I need to do it when the log rendering changes. But we can remove it if you feel it would be cleaner or better deleted 🙇

babakks replied:
Fair. Let's keep them.

BagToad replied:
I edited this comment to better indicate that's what it is for:

```Go
			// Handy note for updating the testdata files when they change:
			// ext := filepath.Ext(tt.log)
			// stripped := strings.TrimSuffix(tt.log, ext)
			// os.WriteFile(stripped+".want"+ext, stdout.Bytes(), 0644)
```

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -240,6 +241,35 @@ func (c *CAPIClient) GetSession(ctx context.Context, id string) (*Session, error
 }
 
+// GetSession retrieves logs of an agent session identified by ID.
+func (c *CAPIClient) GetSessionLogs(ctx context.Context, id string) ([]byte, error) {
```
I forgot to fix the method in godoc.

babakks reviewed (commented):
Thanks, @BagToad! I love the way the logs are displayed now! 🥳 

Just a few notes before marking this as ready

BagToad committed:
Add and improve function documentation in log.go

Added and expanded Go doc comments for multiple functions in pkg/cmd/agent-task/shared/log.go to clarify their purpose and usage. Also refactored renderToolCallTitle and related logic for improved naming consistency and robustness.

BagToad committed:
Fix comment for GetSessionLogs method

BagToad committed:
Make log tests OS-agnostic by normalizing line endings

Updated TestFollow to normalize line endings in test log files, ensuring consistent behavior across different operating systems, especially Windows.

BagToad committed:
Add note for updating testdata files in log tests

A comment was added to TestFollow with instructions on how to update the .want testdata files when test outputs change. This helps maintainers regenerate expected output files more easily.

BagToad committed:
Fix line ending conversion in tests

BagToad committed:
Normalize line endings in log testdata files too

BagToad committed:
Rename testdata log files for clarity

Renamed sample log input and expected output files in testdata to use a consistent 'log-*-input.txt' and 'log-*-want.txt' naming scheme. Updated references in log_test.go to match the new file names for improved clarity and maintainability.

BagToad committed:
Enable Bash session tool handling in log rendering

Uncomments and activates support for Bash session tool calls (write_bash, read_bash, stop_bash, async_bash, read_async_bash, stop_async_bash) in the renderLogEntry function. Also defines the corresponding argument structs, enabling proper handling and display of these tool calls in the log output.

BagToad committed:
Update comment for reasoning text formatting

Clarifies that reasoning text should be formatted as a normal 'thought' message without a heading, improving code documentation.

BagToad committed:
Add TODO for bash-related tool call details

Inserted a TODO comment to consider including more details for bash-related tool calls in the log rendering function.

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+
+					contentWithCommand := choice.Delta.Content
+					if v.Command != "" {
+						contentWithCommand = fmt.Sprintf("%s\n%s", v.Command, choice.Delta.Content)
```
In the GUI, Bash commands are prepended with `$`. Do we want to keep that in our logs? I'd say yes, but want to hear your thoughts.

<img width="422" height="186" alt="Image" src="https://github.com/user-attachments/assets/aa7ddf76-a136-48f5-9516-c3d0f5afda50" />

BagToad replied:
Oh! I missed that. Yeah I don't see why not 😁

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+
+				content := stripDiffFormat(choice.Delta.Content)
+
+				// TODO: Strip the diff formatting from this, but for now render as it is.
+				if err := renderFileContentAsMarkdown(args.Path, content, w, io); err != nil {
+					return false, fmt.Errorf("failed to render viewed file content: %w", err)
+				}
```
I'm not sure if it's okay to bail out if the Markdown renderer fails, because I think it's quite possible for AI to produce flawed Markdown. In such cases, I think we should just ignore the error and continue processing logs.

One more reason is, if we bail out at this point, then the user would never be able to see their session logs, because again we reach to the same point and abort.

BagToad replied:
Makes sense 👍

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+					if v.Command != "" {
+						contentWithCommand = fmt.Sprintf("%s\n%s", v.Command, choice.Delta.Content)
+					}
+					if err := renderFileContentAsMarkdown("commands.sh", contentWithCommand, w, io); err != nil {
+						return false, fmt.Errorf("failed to render bash command output: %w", err)
+					}
```
The same issue.

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+
+				// NOTE: omit the delta.content since it's the same as thought
+				renderToolCallTitle(w, cs, "Thought", "")
+				if err := renderRawMarkdown(args.Thought, w, io); err != nil {
+					return false, fmt.Errorf("failed to render thought: %w", err)
+				}
```
The same. I won't comment on further Markdown renderer cases.

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+				}
+			case "report_progress":
+				args := reportProgressToolArgs{}
+				if err := json.Unmarshal([]byte(tc.Function.Arguments), &args); err != nil {
+					return false, fmt.Errorf("failed to parse 'report_progress' tool call arguments: %w", err)
+				}
```
Basically the same concern. Given the high rate of changes happening everyday, I think we should just be ready to print a generic record when we can't parse the JSON args. There are also other cases down here.

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+		}
+	}
+
+	// If we found the hunk header end, we strip everything before it.
+	if hunkEndIndex != -1 {
+		lines = lines[hunkEndIndex+1:]
+	} else {
+		// This isn't a diff, so we defensively just return the original string.
+		return diff
+	}
```
**nitpick:** This can be a nice guard clause:


```suggestion
	if hunkEndIndex == -1 {
		// This isn't a diff, so we defensively just return the original string.
		return diff
	}
	
	// If we found the hunk header end, we strip everything before it.
	lines = lines[hunkEndIndex+1:]
```

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+	// these prefixes.
+	var stripped []string
+	for _, line := range lines {
+		if strings.HasPrefix(line, "+") || strings.HasPrefix(line, "-") {
+			stripped = append(stripped, line[1:])
+		} else {
+			stripped = append(stripped, line)
+		}
```
Maybe I'm wrong, but lines without `+` or `-` indicate unchanged lines, but they're actually being prefixed with a whitespace. So, if we don't want to misalign the lines I think we should still strip the first char. What do you think?

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+// renderFileContentAsMarkdown renders the given content as markdown
+// based on the file extension of the path.
+func renderFileContentAsMarkdown(path, content string, w io.Writer, io *iostreams.IOStreams) error {
+	parts := strings.Split(path, ".")
+	lang := parts[len(parts)-1]
```
Since `path` can include dots other than the actual file extension, I think we should use the `filepath` package to help us extract the right extension.


```suggestion
	// We need to apply filepath.ToSlash to make sure the behaviour is the same
	// regardless of the platform.
	lang := filepath.Ext(filepath.ToSlash(path))
```

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+	// Glamour adds leading newlines when content is only a code block,
+	// so we only want to add a trailing newline.
+	formatFunc := func(s string) string {
+		return fmt.Sprintf("%s\n\n", s)
```
We're probably fine (as with all other `gh` commands), but we have to check these `\n` usages on Windows terminals.

babakks commented on pkg/cmd/agent-task/shared/log.go:
```diff
@@ -0,0 +1,570 @@
+		"playwright-browser_resize":           "Resize Playwright web browser window",
+		"playwright-browser_close":            "Close Playwright web browser",
+		"playwright-browser_press_key":        "Press key in Playwright web browser",
+		"playwright-browser_select_option":    "Select option in Playwright web browser",
+		"playwright-browser_handle_dialog":    "Interact with dialog in Playwright web browser",
+		"playwright-browser_console_messages": "Get console messages from Playwright web browser",
+		"playwright-browser_drag":             "Drag mouse between elements in Playwright web browser",
+		"playwright-browser_file_upload":      "Upload file in Playwright web browser",
+		"playwright-browser_hover":            "Hover mouse over element in Playwright web browser",
+		"playwright-browser_network_requests": "Get network requests from Playwright web browser",
+
+		// GitHub MCP server common tools
+		"github-mcp-server-get_file_contents":              "Get file contents from GitHub",
+		"github-mcp-server-get_pull_request":               "Get pull request from GitHub",
+		"github-mcp-server-get_issue":                      "Get issue from GitHub",
+		"github-mcp-server-get_pull_request_files":         "Get pull request changed files from GitHub",
+		"github-mcp-server-list_pull_requests":             "List pull requests on GitHub",
+		"github-mcp-server-list_branches":                  "List branches on GitHub",
+		"github-mcp-server-get_pull_request_diff":          "Get pull request diff from GitHub",
+		"github-mcp-server-get_pull_request_comments":      "Get pull request comments from GitHub",
+		"github-mcp-server-get_commit":                     "Get commit from GitHub",
+		"github-mcp-server-search_repositories":            "Search repositories on GitHub",
+		"github-mcp-server-search_code":                    "Search code on GitHub",
+		"github-mcp-server-get_issue_comments":             "Get issue comments from GitHub",
+		"github-mcp-server-list_issues":                    "List issues on GitHub",
+		"github-mcp-server-search_pull_requests":           "Search pull requests on GitHub",
+		"github-mcp-server-list_commits":                   "List commits on GitHub",
+		"github-mcp-server-get_pull_request_status":        "Get pull request status from GitHub",
+		"github-mcp-server-search_issues":                  "Search issues on GitHub",
+		"github-mcp-server-get_pull_request_reviews":       "Get pull request reviews from GitHub",
+		"github-mcp-server-download_workflow_run_artifact": "Download GitHub Actions workflow run artifact",
+		"github-mcp-server-get_job_logs":                   "Get GitHub Actions job logs",
+		"github-mcp-server-get_workflow_run":               "Get GitHub Actions workflow run",
+		"github-mcp-server-get_workflow_run_logs":          "Get GitHub Actions workflow run logs",
+		"github-mcp-server-get_workflow_run_usage":         "Get GitHub Actions workflow usage",
+		"github-mcp-server-list_workflow_jobs":             "List GitHub Actions workflow jobs",
+		"github-mcp-server-list_workflow_run_artifacts":    "List GitHub Actions workflow run artifacts",
+		"github-mcp-server-list_workflow_runs":             "List GitHub Actions workflow runs",
+		"github-mcp-server-list_workflows":                 "List GitHub Actions workflows",
+	}
```
This should be a top-level var.

babakks commented on pkg/cmd/agent-task/shared/log_test.go:
```diff
@@ -0,0 +1,72 @@
+			raw, err := os.ReadFile(tt.log)
+			require.NoError(t, err)
+
+			// Delete all the `/r` to make the tests OS-agnostic.
```
Should be `\r`.

babakks reviewed (approved):
Great job, @BagToad! I put a couple of comments to discuss and potentially apply in a follow up.
